services/mlmodel/client_test.go

```diff
@@ -0,0 +1,140 @@
+	defer rpcServer.Stop()
+
+	// context canceled
+	t.Run("Failing client", func(t *testing.T) {
+		cancelCtx, cancel := context.WithCancel(context.Background())
+		cancel()
+		_, err = viamgrpc.Dial(cancelCtx, listener1.Addr().String(), logger)
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err.Error(), test.ShouldContainSubstring, "canceled")
+	})
+
+	// working
+	t.Run("ml model client", func(t *testing.T) {
+		conn, err := viamgrpc.Dial(context.Background(), listener1.Addr().String(), logger)
+		test.That(t, err, test.ShouldBeNil)
+		client := mlmodel.NewClientFromConn(context.Background(), conn, testMLModelServiceName, logger)
+		// Infer Command
+		result, err := client.Infer(context.Background(), inputData)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, len(result), test.ShouldEqual, 4)
+		// decode the map[string]interface{} into a struct
+		temp := struct {
+			NDetections      []int32       `mapstructure:"n_detections"`
+			ConfidenceScores [][]float32   `mapstructure:"confidence_scores"`
+			Labels           [][]int32     `mapstructure:"labels"`
+			Locations        [][][]float32 `mapstructure:"locations"`
+		}{}
+		err = mapstructure.Decode(result, &temp)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, temp.NDetections[0], test.ShouldEqual, 3)
+		test.That(t, len(temp.ConfidenceScores[0]), test.ShouldEqual, 3)
+		test.That(t, len(temp.Labels[0]), test.ShouldEqual, 3)
+		test.That(t, temp.Locations[0][0], test.ShouldResemble, []float32{0.1, 0.4, 0.22, 0.4})
+		// nil data should work too
+		result, err = client.Infer(context.Background(), nil)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, len(result), test.ShouldEqual, 4)
+		// Metadata Command
+		meta, err := client.Metadata(context.Background())
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, meta.ModelName, test.ShouldEqual, "fake_detector")
+		test.That(t, meta.ModelType, test.ShouldEqual, "object_detector")
+		test.That(t, meta.ModelDescription, test.ShouldEqual, "desc")
+		t.Logf("inputs: %v", meta.Inputs)
+		test.That(t, len(meta.Inputs), test.ShouldEqual, 1)
+		test.That(t, len(meta.Outputs), test.ShouldEqual, 4)
+		outInfo := meta.Outputs
+		test.That(t, outInfo[0].Name, test.ShouldEqual, "n_detections")
+		test.That(t, len(outInfo[0].AssociatedFiles), test.ShouldEqual, 0)
+		test.That(t, outInfo[2].Name, test.ShouldEqual, "labels")
+		test.That(t, len(outInfo[2].AssociatedFiles), test.ShouldEqual, 1)
+		test.That(t, outInfo[2].AssociatedFiles[0].Name, test.ShouldEqual, "category_labels.txt")
+		test.That(t, outInfo[2].AssociatedFiles[0].LabelType, test.ShouldEqual, mlmodel.LabelTypeTensorValue)
+		test.That(t, outInfo[3].Name, test.ShouldEqual, "locations")
+		test.That(t, outInfo[3].NDim, test.ShouldEqual, 3)
+
+		// close the client
+		test.That(t, utils.TryClose(context.Background(), client), test.ShouldBeNil)
+		test.That(t, conn.Close(), test.ShouldBeNil)
+	})
+}
+
+func TestClientDialerOption(t *testing.T) {
```

Comment:
I was on the bandwagon, every other service has this test. 
But it shows the the `rpc.ClientConn` gets cached. 
In https://github.com/viamrobotics/goutils/blob/main/rpc/dialer.go
```
// NewCachedDialer returns a Dialer that returns the same connection if it
// already has been established at a particular target (regardless of the
// options used).
```

`conn1, err := viamgrpc.Dial(ctx, listener.Addr().String(), logger)` creates a new connection with the server, and you store that connection in the client struct when you call `mlmodel.NewClientFromConn`. You can see that it made new connections since `td.NewConnections == 3`

Then you create a new connection, and a new client from that connection, but you demonstrate that the connection was cahced as expected, because it is still `td.NewConnections == 3`.

Reply:
The "Options" being tested are we can store the NewCachedDialer() in the context by using rpc.ContextWithDialer, and the options with the context actually get used.

Reply:
Ah, thanks for explaining!